docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,453 @@
+including Conul, are brought up in a secure configuration.
+It is impossible,
+once certain services have been started (such as Consul),
+to go back and secure them on-the-fly.
+
+Secure bootstrapping is needed for any installation of EdgeX,
+but the problem is more profound for a container-based deployment of EdgeX
+because a docker-compose script does not have an explicit installation phase.
+The information in this ADR is generally applicable to the install phase of an EdgeX snap,
+or any manual installation of EdgeX.
+
+A security-disabled installation of EdgeX is not required to follow this guidance.
+
+### Items requiring secure bootstrapping
+
+#### Secrets seed
+
+The secrets seed is a hardware-protected secret value that is used for key derivation.
+The **seed** can be used as input key material to a key deriviation function
+that generates an unlimited number of derived keys from an initial secret,
+all of which appear to be random.
+
+Derived keys can be used to protect bootstrapping secrets
+that cannot otherwise be protected by the EdgeX secret store.
+
+#### Consul
+
+Consul is a dynamic configuration database that is initially seeded from static configuration files.
+Putting configuration "online" exposes it to unauthorized modification and disclosure
+and requires introducing compensating controls.
+Most at risk is service location:
+information regarding the host and port where an EdgeX service is available.
+To a lesser extent,
+Consul holds general service configuration information
+and is the fallback location for the EdgeX secret store if Vault is not used.
+Consul also provides for service health checks.
+
+The elements required to secure Consul are:
+
+- For neighbor discovery: An **encryption key** used by Serf;
+  this key must be provided the first time that Consul is started.
+  This key prevents rogue Consul instances from joining the cluster.
+  This key must also be provided to any new Consul agents that are added to the system.
+  This key can be a 32-byte derived key or a random key.
+- For TLS: a **CA certificate**, **TLS certificate (chain)**, and **TLS private key**.
+
+For integration with Vault,
+some initial secrets must be created:
+
+- A **management token** that can be passed to Vault's Consul secrets engine,
+  to enable a Vault API to be utilized to create additional per-microservice Consul tokens.
+- An **ACL token** that allows Vault to read/write a portion of Consul's key space as the backing store for Vault.
+  (Intent is to switch to the Vault raft storage backend to avoid Consul dependency.)
+
+#### EdgeX secret store
+
+The EdgeX secret store is based on Hashicorp Vault.
+Upon initialization, Vault generates a **master key**,
+which is used to encrypt the secrets database.
+The master key is used to unseal Vault secrets,
+and to regenerate the Vault root token.
+
+#### EdgeX public key infrastructure
+
+A cluster-wide PKI is needed
```

Comment:
does cluster hint running Vault in high availability mode?

Reply:
No, but it does hint that some of these changes will make distributing EdgeX services easier.  Like switching to raft instead of consul as the Vault storage backend.